File: plugin/graph/src/main/groovy/net/twisterrob/gradle/graph/javafx/interop/TaskDataSerializer.java
package net.twisterrob.gradle.graph.javafx.interop;

import java.lang.reflect.Type;

import com.google.gson.*;

import net.twisterrob.gradle.graph.TaskData;

public class TaskDataSerializer implements JsonSerializer<TaskData> {
	@Override public JsonElement serialize(TaskData data, Type type, JsonSerializationContext context) {
		JsonObject obj = new JsonObject();
		obj.add("label", new JsonPrimitive(data.getTask().getPath()));
		obj.add("type", context.serialize(data.getType()));
		obj.add("state", context.serialize(data.getState()));

		JsonArray deps = new JsonArray();
		for (TaskData dep : data.getDepsDirect()) {
			deps.add(context.serialize(dep.getTask()));
		}
		obj.add("deps", deps);

		return obj;
	}
}

File: graph/src/main/java/net/twisterrob/gradle/graph/vis/d3/interop/TaskDataSerializer.kt
package net.twisterrob.gradle.graph.vis.d3.interop

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import net.twisterrob.gradle.graph.tasks.TaskData
import java.lang.reflect.Type

class TaskDataSerializer : JsonSerializer<TaskData> {

	override fun serialize(data: TaskData, type: Type, context: JsonSerializationContext): JsonElement =
		JsonObject().apply {
			add("label", context.serialize(data.task))
			add("type", context.serialize(data.type))
			add("state", context.serialize(data.state))
			add("deps", data.depsDirect.mapTo { context.serialize(it.task) })
		}
}

inline fun <T> Iterable<T>.mapTo(target: JsonArray = JsonArray(), transform: (T) -> JsonElement): JsonArray {
	for (item in this) {
		target.add(transform(item))
	}
	return target
}
